### PR TITLE
Support promoting multiple users at once in rooms

### PR DESCRIPTION
**Pull request by AnnikaCodes**
Suggested [here](https://www.smogon.com/forums/threads/allow-multiple-arguments-for-promotions-and-demotions.3670212/#post-8592616).

### CODE REVIEW

**Comment by Zarel** [comment]
This function has definitely gotten pretty messy. I wonder if there's a way to refactor.

File: server/chat-commands/moderation.ts [hunk 1]
```diff
@@ -110,76 +92,103 @@ export const commands: ChatCommands = {
...
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting from ${currentGroupName}.`);
+					continue;
+				}
+				if (nextSymbol !== ' ' && !user.can(`room${nextGroup.id || 'voice'}` as 'roomvoice', null, room)) {
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting to ${nextGroupName}.`);
+					continue;
+				}
+			}
+			if (targetUser?.locked && room.persist && room.settings.isPrivate !== true && nextGroup.rank > 2) {
+				this.errorReply("Locked users can't be promoted.");
```
Comment (Zarel) [inline server/chat-commands/moderation.ts:136]: These three error messages (access denied promoting to, access denied promoting from, and locked users) happen every loop and so should mention the username.

File: server/chat-commands/moderation.ts [hunk 2]
```diff
@@ -110,76 +92,103 @@ export const commands: ChatCommands = {
...
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting ${name} from ${currentGroupName}.`);
+					continue;
+				}
+				if (nextSymbol !== ' ' && !user.can(`room${nextGroup.id || 'voice'}` as 'roomvoice', null, room)) {
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting ${name} to ${nextGroupName}.`);
+					continue;
+				}
+			}
+			if (targetUser?.locked && room.persist && room.settings.isPrivate !== true && nextGroup.rank > 2) {
+				this.errorReply(`${name} is locked and therefore can't be promoted.`);
```
Comment (Zarel) [inline server/chat-commands/moderation.ts:136]:
```suggestion
				this.errorReply(`${name} is locked and can't be promoted.`);
```

**Comment by Zarel** [comment]
It looks like most of this command is I/O which isn't easy to refactor. `Chat.ErrorMessage` does make it possible to at least refactor permission checking somewhere nicer, though.

**Comment by AnnikaCodes** [comment]
> It looks like most of this command is I/O which isn't easy to refactor. `Chat.ErrorMessage` does make it possible to at least refactor permission checking somewhere nicer, though.

yeah I'll do that when I'm not in class

**Comment by Zarel** [comment]
Looks good, thanks!

**Pull request merged**